### PR TITLE
[Review]Repartition before shuffle if the requested partitions are less than input partitions 

### PR DESCRIPTION
**Pull request by VibhuJawa**
This PR changes shuffle behavior to repartition before shuffling if requested output partitions is less than input partitions. 
This helps speed up repartitioning quite a bit. 

##### Code
```python
df = dask.datasets.timeseries(end='2010-01-31',partition_freq='12H').persist()
wait(df);

%%time
x = df.shuffle(on=['id'], npartitions=100)
print(len(x))
```
###### Results
| Granularity 	| Input partitions 	| Output Partitions 	| PR 	| Mainline 	| Speed up 	|
|-	|-	|-	|-	|-	|-	|
| 12 H 	| 7366 	| 100 	| 44.3 	| 130 	| 2.93 	|
| 6 H 	| 14732 	| 100 	| 48.3 	| 238 	| 4.93 	|
| 3H 	| 29464 	| 100 	| 55.4 	| 515 	| 9.30 	|

###### PR
Screen Shot 2021-05-26 at 3 56 55 PM

###### Mainline 
Screen Shot 2021-05-26 at 3 56 30 PM
 

### TODO:
- [x] Tests added / passed
- [x] Passes `black dask` / `flake8 dask` / `isort dask`

CC: @rjzamora

### CODE REVIEW

File: dask/dataframe/tests/test_shuffle.py
Comment (rjzamora) [inline dask/dataframe/tests/test_shuffle.py:87]: Any reason for the explicit `scheduler="sync"` here? Can we remove it?

**Review by rjzamora — APPROVED** [review]
Thanks @VibhuJawa! This seems like a great idea to me :)

File: dask/dataframe/tests/test_shuffle.py
Comment (VibhuJawa) [reply dask/dataframe/tests/test_shuffle.py:87]: I just followed the other tests. No particular reason.  We can probably remove it   


https://github.com/dask/dask/blob/2b981ccf90c81997ebcc1a5f9576a7acb184b11a/dask/dataframe/tests/test_shuffle.py#L70-L74

Comment (rjzamora) [reply dask/dataframe/tests/test_shuffle.py:87]: That makes sense.  We probably shouldn't require the single-threaded scheduler in any tests.  I'm sure every instance of that is a leftover debug statement, or someone doing exactly what you did :)  I know I'm guilty of both reasons.

Comment (VibhuJawa) [reply dask/dataframe/tests/test_shuffle.py]: Makes sense. Removed from both with commit https://github.com/dask/dask/pull/7715/commits/86e6deac851cd702c106ba143460f0d5c4c2ce89

Comment (VibhuJawa) [reply dask/dataframe/tests/test_shuffle.py]: Let me know if there is anything else missing. Thanks for the quick review.

**Comment by VibhuJawa** [comment]
@rjzamora, Is this good to get merged in?

**Comment by rjzamora** [comment]
I'm personally happy with this change, but someone else with write access will also need to review :) (cc @dask/maintenance)

**Comment by martindurant** [comment]
So simple! I cannot immediately see why this might be a bad idea. I'll merge tomorrow (when I might be more alert) if no one else objects.

**Pull request merged**
